File: src/Extensibility/SharpDetect.Plugins/Descriptors/MethodVersionDescriptor.cs
// Copyright 2025 Andrej Čižmárik and Contributors
// SPDX-License-Identifier: Apache-2.0

namespace SharpDetect.Plugins.Descriptors;

public record MethodVersionDescriptor(
    int FromMajorVersion,
    int FromMinorVersion,
    int FromBuildVersion,
    int ToMajorVersion,
    int ToMinorVersion,
    int ToBuildVersion)
{
    public static MethodVersionDescriptor Create(Version from, Version to)
    {
        return new MethodVersionDescriptor(
            from.Major,
            from.Minor,
            from.Build,
            to.Major,
            to.Minor,
            to.Build);
    }
}

File: src/SharpDetect.Profiler/SharpDetect.Concurrency.Profiler/MethodVersionDescriptor.cpp
// Copyright 2026 Andrej Čižmárik and Contributors
// SPDX-License-Identifier: Apache-2.0

#include "MethodVersionDescriptor.h"

void Profiler::to_json(nlohmann::json &json, const MethodVersionDescriptor &descriptor)
{
	json["fromMajorVersion"] = descriptor.fromMajorVersion;
	json["fromMinorVersion"] = descriptor.fromMinorVersion;
	json["fromBuildVersion"] = descriptor.fromBuildVersion;
	json["toMajorVersion"] = descriptor.toMajorVersion;
	json["toMinorVersion"] = descriptor.toMinorVersion;
	json["toBuildVersion"] = descriptor.toBuildVersion;
}

void Profiler::from_json(const nlohmann::json &json, MethodVersionDescriptor &descriptor)
{
	descriptor.fromMajorVersion = json.at("fromMajorVersion");
	descriptor.fromMinorVersion = json.at("fromMinorVersion");
	descriptor.fromBuildVersion = json.at("fromBuildVersion");
	descriptor.toMajorVersion = json.at("toMajorVersion");
	descriptor.toMinorVersion = json.at("toMinorVersion");
	descriptor.toBuildVersion = json.at("toBuildVersion");
}